service/history/replication/task_ack_manager.go

```diff
@@ -97,109 +97,117 @@ func NewTaskAckManager(
 
 		maxReplicationMessagesSize: config.MaxResponseSize,
 		replicationMessagesSizeFn:  replicationMessagesSizeFn,
+		dynamicTaskBatchSizer:      newDynamicTaskBatchSizerImpl(shardID, config),
 	}
 }
 
-func (t *TaskAckManager) GetTasks(ctx context.Context, pollingCluster string, lastReadTaskID int64) (*types.ReplicationMessages, error) {
-	if lastReadTaskID == common.EmptyMessageID {
-		lastReadTaskID = t.ackLevels.GetClusterReplicationLevel(pollingCluster)
+func (t *TaskAckManager) GetTasks(ctx context.Context, pollingCluster string, lastReadTaskID int64) (_ *types.ReplicationMessages, err error) {
+	s := t.newGetTasksState(pollingCluster, lastReadTaskID)
+	defer t.dynamicTaskBatchSizer.analyse(err, s)
+
+	if s.lastReadTaskID == common.EmptyMessageID {
+		s.lastReadTaskID = s.previousReadTaskID
 	}
 
 	taskGeneratedTimer := t.scope.StartTimer(metrics.TaskLatency)
+	defer taskGeneratedTimer.Stop()
 
-	tasks, hasMore, err := t.reader.Read(ctx, lastReadTaskID, t.ackLevels.GetTransferMaxReadLevel())
+	batchSize := t.dynamicTaskBatchSizer.value()
+	t.scope.UpdateGauge(metrics.ReplicationTasksBatchSize, float64(batchSize))
+
+	s.taskInfos, s.msgs.HasMore, err = t.reader.Read(ctx, s.lastReadTaskID, t.ackLevels.GetTransferMaxReadLevel(), batchSize)
```

Comment:
Using a locally scoped struct to keep state looks a little bit indirect and error prone to me. It would be more readable to just define the fields as first class variables here and use them instead of `s.` prefixing throughout the func body. There's no method attached to that struct so not sure about the benefit